content/blog/2022/dependabot-single-pull-request-python.md

```diff
@@ -0,0 +1,132 @@
+---
+title: Replace Dependabot with a single Python dependency upgrade pull request
+author: ed
+date: 2022-05-27
+tags:
+  - Article
+  - Build Tools
+  - DevOps
+  - Pip
+  - Python
+  - Tips
+image:
+  src: blog/2022/bot.jpg
+summary: |
+  Keeping your Python dependencies up to date can become an unwieldy task when
+  Dependabot opens a dozen pull requests every week. We present an alternative
+  approach that uses only one for all upgrades.
+---
+
+[Dependabot](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuring-dependabot-version-updates)
+is a useful tool provided by GitHub to all repositories. By including a simple
+`dependabot.yml` file maintainers get an automatic pull request every time one
+of their dependencies releases a new version. Combined with a robust automated
+test suite this greatly reduces the burden of manually keeping dependencies up
+to date.
+
+However, as projects grow and more dependencies are added the noise produced by
+individual pull requests starts becoming an issue on itself. That's why a
+"grouped updates" feature for Dependabot has been
+[requested](https://github.com/dependabot/dependabot-core/issues/1190) since
+[2018](https://github.com/dependabot/dependabot-core/issues/2265). As of May
+2022, the feature is on [Dependabot's
+roadmap](https://github.com/github/roadmap/issues/148), but doesn't have a
+concrete release date.
+
+At OddBird we decided to leverage [GitHub
+Actions](https://github.com/features/actions) to produce a pull request with all
+Python dependencies updated at once. We have created a workflow that runs every
+Monday in the early morning. By the time we start our work day the pull request
+has been opened and we can merge it right away if all tests pass, or debug any
+issues introduced by upstream breaking changes. We also gave us the option of
+[running the workflow on
+demand](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow)
+to upgrade dependencies at our convenience.
+
+Broadly speaking the workflow will:
+
+1. Checkout our code
+1. Configure a Python environment that matches our project version
+1. Run a custom command to upgrade all dependencies
+1. Use git to detect changed files
+1. Commit the changes to a pre-defined branch (if they exist)
+1. Open a pull request for that branch (if it doesn't exist already)
```

Comment:
That's a markdown trick to avoid having to manually keep numbers in order when inserting or shuffling list items around 😅 [It produces the expected `ol`](https://github.com/oddbird/oddleventy-built/blob/1c18835154625102551c6c76b7cc387839dc4543/2022/05/27/dependabot-single-pull-request-python/index.html#L462-L469) but I'm happy to change it for consistency with other posts if you prefer

Reply:
No, I love it -- just didn't know about the trick!